src/vs/workbench/contrib/debug/browser/debugService.ts

```diff
@@ -701,7 +701,7 @@ export class DebugService implements IDebugService {
 	}
 
 	async restartSession(session: IDebugSession, restartData?: any): Promise<any> {
-		await this.editorService.saveAll();
+		await saveAllBeforeDebugStart(this.configurationService, this.editorService);
```

Comment:
When debugging starts, there is some other logic in deciding whether to save, and we should respect that here. For example, extensions, can set a flag to suppress the save:

https://github.com/microsoft/vscode/blob/5bbf85a112e6cd8b024059bc99f6baef2bb43b29/src/vs/workbench/api/browser/mainThreadDebugService.ts#L236-L237

Reply:
I see this in the source code, but I'm looking at the extension API and I'm not seeing the `supressSaveBeforeStart` flag under debug options https://code.visualstudio.com/api/references/vscode-api#DebugSessionOptions

Reply:
That is proposed API: https://github.com/microsoft/vscode/blob/a72212f1f4950dac3bcb7dbfcf805d96adf3c564/src/vscode-dts/vscode.proposed.notebookDebugOptions.d.ts#L27

Reply:
Ideally, `surpressSaveBeforeStart` would be an attribute in `IDebugSessionOptions` so that it could be stored in the session here https://github.com/microsoft/vscode/blob/e790fa3b059584b9d7f187ce461c9adae985698d/src/vs/workbench/contrib/debug/browser/debugSession.ts#L47 and there would be a getter for it (`this._options.surpressSaveBeforeStart`) on the `DebugSession` that could be used when we restart. Is that what you were thinking?

Reply:
However, I also see that `startDebugging` currently takes the `supressSaveBeforeStart` info as an extra argument, so is there a reason why these are separate?

Reply:
Yes, I think what you describe is what we need. The original reason that flag wasn't on the options already was just that we considered it a flag that we need for starting the session, not something that has to stick around as a property of the session. But now since we need it later, I guess it is actually a property of the session.